faucet/src/lib.rs

```diff
@@ -56,125 +63,277 @@ pub mod pallet {
+		+ encointer_communities::Config
+		+ pallet_treasury::Config
+	{
 		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
-		type Currency: Currency<Self::AccountId>;
+		type Currency: Currency<Self::AccountId> + NamedReservableCurrency<Self::AccountId>;
 		type ControllerOrigin: EnsureOrigin<Self::RuntimeOrigin>;
-
 		#[pallet::constant]
 		type PalletId: Get<PalletId>;
 	}
 
 	#[pallet::call]
-	impl<T: Config> Pallet<T> {
-		/// Transfer some balance to another account.
+	impl<T: Config> Pallet<T>
+	where
+		sp_core::H256: From<<T as frame_system::Config>::Hash>,
+		T::AccountId: AsRef<[u8; 32]>,
+		ReserveIdentifierOf<T>: From<[u8; 8]>,
+	{
 		#[pallet::call_index(0)]
 		#[pallet::weight(10_000)]
+		pub fn create_faucet(
+			origin: OriginFor<T>,
+			name: FaucetNameType,
+			amount: BalanceOf<T>,
+			whitelist: WhiteListType,
+			drip_amount: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let from = ensure_signed(origin)?;
+
+			let all_communities = encointer_communities::Pallet::<T>::community_identifiers();
+			for cid in &whitelist {
+				if !all_communities.contains(cid) {
+					return Err(<Error<T>>::InvalidCommunityIdentifierInWhitelist.into())
+				}
+			}
+
+			// create account
+			let faucet_identifier =
+				[<T as Config>::PalletId::get().0.as_slice(), name.to_vec().as_slice()].concat();
+
+			let faucet_id_hash: H256 = T::Hashing::hash_of(&faucet_identifier).into();
+			let faucet_account = T::AccountId::decode(&mut faucet_id_hash.as_bytes())
+				.expect("32 bytes can always construct an AccountId32");
+
+			if <Faucets<T>>::contains_key(&faucet_account) {
+				return Err(<Error<T>>::FaucetAlreadyExists.into())
+			}
+
+			<T as Config>::Currency::reserve_named(
+				&Self::get_reserve_id(&faucet_account),
+				&from,
+				Self::reserve_amount(),
+			)?;
+
+			<T as Config>::Currency::transfer(&from, &faucet_account, amount, KeepAlive)
+				.map_err(|_| <Error<T>>::InsuffiecientBalance)?;
+
+			let purpose_id = <encointer_reputation_commitments::Pallet<T>>::do_register_purpose(
+				DescriptorType::try_from(faucet_identifier)
+					.map_err(|_| <Error<T>>::PurposeIdCreationFailed)?,
+			)?;
+
+			<Faucets<T>>::insert(
+				&faucet_account,
+				Faucet { name: name.clone(), purpose_id, whitelist, drip_amount, creator: from },
+			);
+
+			info!(target: LOG, "faucet created: {:?}, {:?}", name, faucet_account);
+			Self::deposit_event(Event::FaucetCreated(faucet_account, name));
+
+			Ok(().into())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(10_000)]
 		pub fn drip(
 			origin: OriginFor<T>,
+			faucet_account: T::AccountId,
 			cid: CommunityIdentifier,
 			cindex: CeremonyIndexType,
 		) -> DispatchResultWithPostInfo {
 			let from = ensure_signed(origin)?;
 
+			if !<Faucets<T>>::contains_key(&faucet_account) {
+				return Err(<Error<T>>::InexsistentFaucet.into())
+			}
+
+			let faucet = Self::faucets(&faucet_account).ok_or(<Error<T>>::InexsistentFaucet)?;
```

Comment:
seems redundant why test first instead of fetching and catching the possible error?